File: iq/1/twosum/solution.go
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package twosum

// Find returns indices of the first pair of items from nn that add up to x.
func Find(nn []int, x int) []int {
	if len(nn) < 2 {
		return nil
	}
	left, right := 0, len(nn)-1
	for left < right {
		sum := nn[left] + nn[right]
		if sum < x {
			left += 1
		} else if sum > x {
			right -= 1
		} else {
			return []int{left, right}
		}
	}
	return nil
}

File: iq/1/twosum/solution.cc
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//
// clang-format off-next-line
//go:build ignore

#include <optional>
#include <utility>
#include <vector>

namespace iq::twosum {
namespace {

constexpr int kSizeMin = 2;

}  // namespace

std::optional<std::pair<int, int>> Find(const std::vector<int>& nn, int x) {
  if (nn.size() < kSizeMin) {
    return std::nullopt;
  }
  int leftidx = 0;
  int rightidx = nn.size() - 1;
  while (leftidx < rightidx) {
    const int sum = nn[leftidx] + nn[rightidx];
    if (sum < x) {
      ++leftidx;
    } else if (sum > x) {
      --rightidx;
    } else {
      return std::make_pair(leftidx, rightidx);
    }
  }
  return std::nullopt;
}

}  // namespace iq::twosum
